src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -0,0 +1,248 @@
+    {
+        foreach ($reflectionClass->getInterfaces() as $interfaceName => $reflectionInterface) {
+            $this->set($interfaceName, $id);
+
+            $this->extractInterfaces($id, $reflectionInterface);
+        }
+    }
+
+    /**
+     * Extracts all inherited types of a class.
+     *
+     * @param string           $id
+     * @param \ReflectionClass $reflectionClass
+     */
+    private function extractAncestors($id, \ReflectionClass $reflectionClass)
+    {
+        $this->set($reflectionClass->name, $id);
+
+        if ($reflectionParentClass = $reflectionClass->getParentClass()) {
+            $this->extractAncestors($id, $reflectionParentClass);
+        }
+    }
+
+    /**
+     * Associates a type and a service id if applicable.
+     *
+     * @param string $type
+     * @param string $id
+     */
+    private function set($type, $id)
+    {
+        if (isset($this->definedTypes[$type]) || isset($this->notGuessableTypes[$type])) {
+            return;
+        }
+
+        if (isset($this->types[$type])) {
+            if ($this->types[$type] === $id) {
+                return;
+            }
+
+            unset($this->types[$type]);
+            $this->notGuessableTypes[$type] = true;
+
+            return;
+        }
+
+        $this->types[$type] = $id;
+    }
+
+    /**
+     * Registers a definition for the type if possible or throws an exception.
+     *
+     * @param \ReflectionClass $typeHint
+     * @param string           $id
+     *
+     * @return Reference A reference to the registered definition
+     *
+     * @throws RuntimeException
+     */
+    private function createAutowiredDefinition(\ReflectionClass $typeHint, $id)
+    {
+        if (!$typeHint->isInstantiable()) {
+            throw new RuntimeException(sprintf('Unable to autowire argument of type "%s" for the service "%s".', $typeHint->name, $id));
+        }
+
+        $argumentId = sprintf('autowired.%s', $typeHint->name);
+
+        $argumentDefinition = $this->container->register($argumentId, $typeHint->name);
+        $argumentDefinition->setPublic(false);
+
+        $this->populateAvailableType($argumentId, $argumentDefinition);
+        $this->completeDefinition($argumentId, $argumentDefinition);
+
+        return new Reference($argumentId);
+    }
+
+    /**
+     * Retrieves the reflection class associated with the given service.
+     *
+     * @param string     $id
+     * @param Definition $definition
+     *
+     * @return \ReflectionClass|null
+     */
+    private function getReflectionClass($id, Definition $definition)
+    {
+        if (isset($this->reflectionClasses[$id])) {
+            return $this->reflectionClasses[$id];
+        }
+
+        if (!$class = $definition->getClass()) {
+            return;
+        }
+
+        $class = $this->container->getParameterBag()->resolveValue($class);
+
+        return $this->reflectionClasses[$id] = new \ReflectionClass($class);
```

Comment:
you should catch ReflectionException for non-existent classes IMO, to avoid breaking the case where bundles define services for their optional deps, relying on the fact that you will install the dependency if you want to use this feature.